### PR TITLE
Stop OIDC flow when remote writing with bad auth header

### PR DESCRIPTION
**Pull request by douglascamata**
When handling remote write requests, if the Authorization
header isn't present, we don't want to redirect users back
to the identity provider for login.

The reason behind this is that tenants can overload the
identity provider in case they have bad configuration.
We want to be friendly and help prevent this issue from
happening.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -520,6 +520,10 @@ func main() {
 
 			r.Use(authentication.WithTenant)
 			r.Use(authentication.WithTenantID(tenantIDs))
+			// We want to enforce that all remote write requests have an access token present in the Authorization
+			// header to avoid repeatedly redirecting clients that might be unaware of OIDC flow to the identity
+			// provider, possibly overloading it.
+			r.Use(authentication.EnforceAccessTokenPresent("/api/v1/receive"))
```
Comment (matej-g) [inline main.go:526]: I'm wondering if we could make this somehow more universal, since right now this will only apply to metrics / remote write, but similar situation could occur with logs and traces as well right?

**Review by matej-g — COMMENTED** [review]
Thanks for the initiative :heart: This approach is looking great to me, I just have few minor points. You would also get extra coolness points if we had a test for this :sunglasses:.

File: authentication/http.go [hunk 2]
```diff
@@ -54,6 +54,30 @@ func WithTenantID(tenantIDs map[string]string) Middleware {
...
+
+			rawToken := r.Header.Get("Authorization")
+			if rawToken == "" {
+				http.Error(w, "couldn't find the authorization header", http.StatusBadRequest)
+				return
+			}
+			if !strings.Contains(rawToken, "Bearer ") {
+				http.Error(w, "malformed authorization header", http.StatusBadRequest)
+				return
+			}
```
[the remark this answers is not in the document]
Comment (PhilipGough) [reply authentication/http.go:75]: I tend to agree with @matej-g here, but just out of interest, does the presence of an `Authorization` header alone avoid the server sending the client into this redirect hotloop or does the situation also occur wit an invalid/expired token?

File: main.go (hunk 1 above)
Comment (douglascamata) [reply main.go:526]: @matej-g I don't know much how writing logs/traces work, but I suspect they are susceptible to the same kind of situation. I'm happy to use this approach for these other telemetry signals in follow up PRs. :)

File: authentication/http.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (douglascamata) [reply authentication/http.go:75]: We could leave it out, at the cost of decreasing protection from tenants "DoSing" the identity provider with bad configuration. 

The idea of doing this very simple check here is exactly to hard-stop a request **before** it reaches the OIDC middleware, because this kind of failure there means a redirect to the identity provider's login page.

What do you think?

File: main.go (hunk 1 above)
Comment (PhilipGough) [reply main.go:526]: We know the write endpoints upfront right? So perhaps renaming the function to something like `EnsureAccessTokenPresentOnWrite()` or similar and have a check against all endpoints?

File: authentication/http.go [hunk 3]
```diff
@@ -54,6 +54,30 @@ func WithTenantID(tenantIDs map[string]string) Middleware {
...
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !strings.HasSuffix(r.URL.Path, urlPath) {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			rawToken := r.Header.Get("Authorization")
+			if rawToken == "" {
+				http.Error(w, "couldn't find the authorization header", http.StatusBadRequest)
```
Comment (PhilipGough) [inline authentication/http.go:69]: Is there potentially any other flows we could break here? cookies etc?

File: authentication/http.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (PhilipGough) [reply authentication/http.go:75]: what about situations other than OIDC, I see support for mTLS and OpenShift authenticator for example? 🤔

[the remark this answers is not in the document]
Comment (douglascamata) [reply authentication/http.go:75]: I don't think Prometheus remote write supports OpenShift authenticator, so this discards the OpenShift authenticator at least, correct?

I am investigating how can we make this not break the mTLS authentication. Sadly there aren't tests for this. :(

[the remark this answers is not in the document]
Comment (matej-g) [reply authentication/http.go:75]: > what about situations other than OIDC, I see support for mTLS and OpenShift authenticator for example? thinking

That's a good point actually, should this be then part of the OIDC middleware rather, as it applies only to tenants with OIDC authentication providers?

[the remark this answers is not in the document]
Comment (douglascamata) [reply authentication/http.go:75]: Thanks for the comments too, @PhilipGough and @esnible. I am taking your suggestions into account. 👍

**Comment by douglascamata** [comment]
Summarizing the recent changes after applying review's feedback:

* Check on content of the `Authorization` header removed.
* No redirect to login in both Prometheus remote write **and** Loki log push endpoints. Traces already don't have the login redirect (it runs on the gRPC middleware stack).
* The core logic is separate from the OIDC authenticator middleware to avoid injecting knowledge about what's behind authentication into the authenticator.

**Comment by douglascamata** [comment]
@matej-g @esnible @PhilipGough looks like the Github feature for re-requesting a PR view is a bit messed up. This is ready for another look from you.

**Comment by douglascamata** [comment]
By the way, I can give a try to adding a test for this in another PR. I am afraid it will take too long and add too much modification to a single PR though, @matej-g

File: authentication/oidc.go [hunk 4]
```diff
@@ -434,3 +434,46 @@ func (a oidcAuthenticator) checkAuth(ctx context.Context, token string) (context
...
 }
+
+// EnforceAccessTokenPresentOnSignalWrite enforces that the Authorization header is present in the incoming request
+// for the given list of tenants. Otherwise, it returns an error.
+// It protects the Prometheus remote write and Loki push endpoints. The tracing endpoint is not protected because
+// it goes through the gRPC middleware stack, which behaves differently from the HTTP one.
+func EnforceAccessTokenPresentOnSignalWrite(oidcTenants map[string]struct{}) Middleware {
+	protectedSignalWritePath := []string{
+		// Prometheus remote write endpoint
+		"/api/v1/receive",
```
Comment (squat) [inline authentication/oidc.go:445]: This looks to me like we are crossing abstraction boundaries. I'm worried about hardcoding API endpoints for data into authentication middleware. The auth middleware should really not know anything about this. Imagine we one day change an endpoint or add a new one, now we have to somehow remember to update this list as well. I would worry this is asking for vulnerability to be introduced.

Maybe we could get around this by having a middleware factory that accepts a list of endpoints and we can produce this middleware in the main.go.

Comment (douglascamata) [reply authentication/oidc.go:445]: > This looks to me like we are crossing abstraction boundaries.
I had a similar feeling about it, @squat. I tried to think of a way to prepend (not append, because we need this to run before the auth middleware, as it will do the redirect) this middleware to the stack of each sub-router (for metrics, logs, and traces... in their respective blocks in `main.go`). I didn't find a good way of doing this though. I'll investigate this more.

I was even wondering whether we could "cut the problem at its root" and remove the login redirect in the OIDC authentication flow and let it a responsibility of the clients. What do you think?

**Comment by jessicalins** [comment]
thanks for adding this!

> By the way, I can give a try to adding a test for this in another PR. I am afraid it will take too long and add too much modification to a single PR though, @matej-g
^ I'd vote for adding a test on this PR too

**Comment by douglascamata** [comment]
To be able to write tests for this using the e2e framework, we need https://github.com/observatorium/up/pull/44 to be merged.

File: authentication/http.go [hunk 5]
```diff
@@ -158,3 +158,42 @@ func WithTenantMiddlewares(mwFns ...MiddlewareFunc) Middleware {
...
+			if _, found := oidcTenants[tenant]; !found {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			// We aren't interested in blocking requests from tenants not using OIDC.
+			if _, found := oidcTenants[tenant]; !found {
+				next.ServeHTTP(w, r)
+				return
+			}
```
Comment (matej-g) [inline authentication/http.go:187]: Looks like this is an erroneous duplicate with lines 178-181?

**Review by matej-g — COMMENTED** [review]
Great job @douglascamata and thanks for adding tests! I'm totally happy with current state, minus removing the duplicate code lines. After that all :green_circle:.

File: authentication/http.go (hunk 5 above)
Comment (douglascamata) [reply authentication/http.go]: Totally! It must have passed through some merge conflict or something. Fixed in a3de8f4. Thanks!

**Review by matej-g — APPROVED** [review]

**Review by matej-g — DISMISSED** [review]

File: main.go [hunk 6]
```diff
@@ -8,6 +8,7 @@ import (
 	"errors"
 	"flag"
 	"fmt"
+	"github.com/metalmatze/signal/internalserver"
```
Comment (PhilipGough) [inline main.go:11]: tiny nit: This is an unnecessary diff

**Review by PhilipGough — APPROVED** [review]
looks good @douglascamata great job!!

File: api/metrics/legacy/http.go [hunk 7]
```diff
@@ -71,6 +71,14 @@ func WithUIMiddleware(m func(http.Handler) http.Handler) HandlerOption {
 	}
 }
 
+// WithGlobalMiddleware adds a middleware for all operations.
+func WithGlobalMiddleware(m ...func(http.Handler) http.Handler) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.uiMiddlewares = append(h.uiMiddlewares, m...)
+		h.queryMiddlewares = append(h.queryMiddlewares, m...)
+	}
```
Comment (PhilipGough) [inline api/metrics/legacy/http.go:79]: do we need to update the legacy path?

Comment (douglascamata) [reply api/metrics/legacy/http.go:79]: The legacy path doesn't seem to accept remote writes, so we don't need to update it. Serving just UI and query it's "safe".

File: main.go (hunk 6 above)
Comment (douglascamata) [reply main.go:11]: Auto import 😅

Comment (douglascamata) [reply main.go]: Fixed in [4c521e0](https://github.com/observatorium/api/pull/293/commits/4c521e0d51cbae0f67e1ed6343414bc359a142d3)

**Comment by douglascamata** [comment]
@PhilipGough: can you please re-review the PR? I fixed your last comment. :)

**Review by PhilipGough — APPROVED** [review]
